.github/workflows/ci-tests.yml

```diff
@@ -69,35 +81,129 @@ jobs:
-        with:
-          command: test
+      - name: Clean unused crate source checkouts and git repo checkouts
+        run: cargo cache
 
-      - name: cargo clippy --locked (Linux & Windows)
-        uses: actions-rs/clippy-check@b5b5f21f4797c02da247df37026fcd0a5024aa4d # @v1.0.7
+      - name: cargo clippy (Linux & Windows)
+        uses: actions-rs/clippy-check@b5b5f21f4797c02da247df37026fcd0a5024aa4d # v1.0.7
         with:
           token: ${{ secrets.GITHUB_TOKEN }}
           args: --locked --all-targets -- -D warnings
         if: runner.os != 'macOS'
 
-      - name: cargo clippy --locked (macOS)
+      - name: cargo clippy (MacOS)
         uses: actions-rs/clippy-check@b5b5f21f4797c02da247df37026fcd0a5024aa4d # @v1.0.7
         with:
           token: ${{ secrets.GITHUB_TOKEN }}
-          args: --locked --all-targets --no-default-features -- -D warnings
+          args: --locked  --no-default-features -- -D warnings
         if: runner.os == 'macOS'
 
+      - name: Clean unused artifacts
+        run: cargo sweep --maxsize ${{ env.MAX_TARGET_SIZE }}
+
+  test:
+    strategy:
+      matrix:
+        os: ${{ fromJson(github.repository_owner == 'subspace' && '[["self-hosted", "ubuntu-20.04-x86-64"], ["self-hosted", "macos-12-arm64"], ["self-hosted", "windows-server-2022-x86-64"]]' || '["ubuntu-20.04", "macos-12", "windows-2022"]') }}
+        run-all:
+          - ${{ inputs.test-macos-and-windows || github.ref == 'refs/heads/master' }}
+        exclude: # exclude macos-12 and window-2022 when the condition is false
+          - run-all: false
+            os: macos-12
+          - run-all: false
+            os: windows-2022
+
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: git checkout
+        uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # v3.1.0
+
+      # On macOS, we need a proper Clang version, not Apple's custom version without wasm32 support
+      - name: Install LLVM and Clang
+        uses: KyleMayes/install-llvm-action@8852e4d5c58653ed05135c0a5d949d9c2febcb00 # v1.6.1
+        with:
+          version: "15.0"
+        if: runner.os == 'macOS'
+
+      - name: Install Protoc
+        uses: arduino/setup-protoc@9b1ee5b22b0a3f1feb8c2ff99b32c89b3c3191e9 # v2.0.0
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+
+      # Needed for hwloc
+      - name: Install automake (macOS)
+        run: brew install automake
+        if: runner.os == 'macOS'
+
+      # Workaround to resolve link error with C:\msys64\mingw64\bin\libclang.dll
+      - name: Remove msys64
+        run: Remove-Item -LiteralPath "C:\msys64\" -Force -Recurse
+        if: runner.os == 'Windows'
+        continue-on-error: true
+
+      - name: Add cache
+        uses: actions/cache@704facf57e6136b1bc63b828d79edcd491f0ee84 # @v3.3.2
+        if: runner.os != 'Windows'
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/bin
+            ~/.cargo/git
+            target
```

Comment:
I don't think caching `target` is a good idea. It is not recommended and will not fit into cache anyway due to size limit. And even if it does fit, it'll quickly get evicted from cache due to large total size.